### PR TITLE
[InstCombine] Add folds for (icmp eq/ne (and (add/sub/xor A, P2), P2), 0/P2)

### PR DESCRIPTION
**Pull request by goldsteinn**
- [InstCombine] Add tests for folding (icmp eq/ne (and (add/sub/xor A, P2), P2), 0/P2); NFC
- [InstCombine] Add folds for (icmp eq/ne (and (add/sub/xor A, P2), P2), 0/P2)

All associated proofs: https://alive2.llvm.org/ce/z/45pq2x

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -5453,6 +5453,45 @@ Instruction *InstCombinerImpl::foldICmpEquality(ICmpInst &I) {
...
+      IsZero = MatchAndP2OfAddSubXor(Instruction::Sub);
+    if (!IsZero)
+      IsZero = MatchAndP2OfAddSubXor(Instruction::Xor);
+
+    if (IsZero) {
+      // (icmp eq/ne (and (add/sub/xor X, P2), P2), P2)
+      //    -> (icmp eq/ne (and X, P2), 0)
+      // (icmp eq/ne (and (add/sub/xor X, P2), P2), 0)
+      //    -> (icmp eq/ne (and X, P2), P2)
+      return new ICmpInst(Pred, Builder.CreateAnd(A, B),
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5489]:
```suggestion
      return new ICmpInst(Pred, Builder.CreateAnd(B, A),
```
A is likely to be a constant.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -5453,6 +5453,45 @@ Instruction *InstCombinerImpl::foldICmpEquality(ICmpInst &I) {
...
+                                    m_Value(A),
+                                    m_c_BinOp(Opc, m_Value(B), m_Deferred(A)))),
+                                m_Zero())))
+        Matched = true;
+
+      if (Matched && isKnownToBeAPowerOfTwo(A, /* OrZero */ true, 0, &I))
+        return Matched;
+      return std::nullopt;
+    };
+    std::optional<bool> IsZero = MatchAndP2OfAddSubXor(Instruction::Add);
```
[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5478]: Only for constants and only if the add/sub/xor has one use. This fold is trying to address the other case.

[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5479]: It is easy to do canonicalization for non-constant.
Alive2: https://alive2.llvm.org/ce/z/LmLn2R
I will submit a patch later.

[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5479]: Posted as #67915.

[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5479]: In favor of your commit but still want this for multiuse case. Although realize there is bug in this code b.c commutative sub.

[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Fixed the bug + cleaned up codes (your pattern is much cleaner) + made tests more explicit about multi-use case.

**Comment by dtcxzyw** [comment]
#67915 has been merged. Please update pre-commit tests.

**Comment by goldsteinn** [comment]
> #67915 has been merged. Please update pre-commit tests.

No diff.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Comment by goldsteinn** [comment]
ping.

**Comment by nikic** [comment]
Just to double check: Is there real world motivation for this pattern? As this is a multi-use-only pattern, we should not handle it without specific motivation. (As you can imagine, you can construct a huge number of multi-use-only patterns by combining any number of use-restricted patterns.)

**Comment by goldsteinn** [comment]
> Just to double check: Is there real world motivation for this pattern? As this is a multi-use-only pattern, we should not handle it without specific motivation. (As you can imagine, you can construct a huge number of multi-use-only patterns by combining any number of use-restricted patterns.)

Motivation case was essentially odd-even pattern where noticed the cmp was happen after the increment/decrement. The `sub`/`xor` + `P2` form was just to cover the general pattern.

**Comment by antoniofrighetto** [comment]
Just FYI, I counted overall more improvements than regressions when disabling this folding (https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1726).

**Pull request closed without merging**